### ISSUE TITLE
[engine] Engine should automatically detect when classes and config are changed and reload the context

### ISSUE THREAD

**Issue by russdanner** [craftercms/craftercms#6223]
### Latest version

- [X] The issue is in the develop branch
- [X] The issue is in the latest released 4.1.x
- [X] The issue is in the latest released 4.0.x
- [X] The issue is in the latest released 3.1.x

### Duplicates

- [X] I have searched the existing issues

### Is your enhancement request related to a problem? Please describe.

Today changes to nested groovy classes often require a context rebuild in order to get picked up. Ideally, CrafterCMS could detect a development change that would require a context rebuild and handle that seamlessly for the developer.

Example:
A rest controller imports a class written in groovy and uses it. The user changes the class but not the rest script. The updated class changes are not seen until after a context rebuild (or a change to the rest controller)

### Describe the solution you'd like

Engine should automatically detect when classes and config are changed and reload the context.

**Comment by phuongnq** [comment]
@sumerjabri @russdanner The implementation enables the DirectoryWatcher for `/scripts` and `/site/engine` and for preview mode.
Should we have a separate flag to enable this instead of preview mode?

**Comment by sumerjabri** [comment]
@phuongnq the Deployer should not rebuild context for Preview targets now that we're doing that in Engine. See if that's possible, please.

**Comment by sumerjabri** [comment]
Suggested test cases:
* Write/change/delete files to disk (no commit) in the monitored paths: /scripts/classes/.../*.java (a class file) -- see if Engine picks it up and it runs with the new code
* Same for config files (Spring context updates, etc.)
* Git pull similar files from remote where these files are updated

**Comment by phuongnq** [comment]
@sumerjabri Regarding the Deployer, since we use the `local-target-template.yaml` for the preview. I have removed the http processor to do the rebuild. (Assuming we use this template for preview only?).

Those test cases are tested in Linux. For the pull, since they will have the same modified date, I updated the logic to do just one rebuild instead of multiple rebuilds for each file since one rebuild would be enough. If the modified date are different for multiple files, we will treat it as multiple rebuild requests.

We may also want to test the same in MacOS since the library handle watcher differently in the MacOS.

**Comment by sumerjabri** [comment]
We're okay with the Deployer template.

The main issue is the algorithm. My suggested algorithm is:
Main idea: Trigger once for a large set of changes, but never wait more than 1 second
Ingredients:
- A counter with a configurable limit (default 5)
- A configurable timeout (default 200 milliseconds)

Alg:
- When the watcher triggers, set a counter to 1 and sleep for 200 milliseconds.
- Wake up and check:
  - Is the counter 5 (one second has passed)? Then, trigger a rebuild.
  - Has anything else changed (more changes since we slept)? If so, increment the counter and sleep for 200 milliseconds. If nothing has changed, then trigger a rebuild.
- All of the above happens in the same thread, or a thread pool with limits is used.

Let me know your thoughts. I can also meet with you as needed.

**Comment by sumerjabri** [comment]
@yacdaniel you may start testing in `develop`

@phuongnq please port to 4.0

**Comment by Euquimides** [comment]
Hi @sumerjabri and @phuongnq ! While validating the ticket I noticed the following. Changing the configuration of the new engine watcher to watch paths in general like:

```

**Comment by phuongnq** [comment]
@sumerjabri We can have a list of ignore path patterns such as:

```
# Engine file change watcher ignores path patterns
crafter.engine.watcher.ignorePathPatterns=\
  /.git/**
```
Please let me know if we want to implement this. However, in a normal case, we should only rebuild on the default value including the `/config/engine` and `/scripts` folder.

**Comment by sumerjabri** [comment]
@phuongnq let's build the ignore paths feature and default it to `/.git` as you indicated. It would be great if we can do the logic like this:
* Compute what we are asked to watch
* Subtract what we are asked to ignore
This way we can have it exclude specific files even in `/config`

**Comment by Euquimides** [comment]
Hi! Fix verified for Crafter v4.0.8 and v.4.1.2. Thank you! Closing this ticket.
